lambda/trmc.ml

```diff
@@ -0,0 +1,515 @@
+      Lprim (Pmakeblock (tag, new_flag, shape), new_block_args, loc) in
+    let choices = List.map (choice ctx) blockargs in
+    match find_settable choices with
+    | Error all_returns -> Return (k flag all_returns)
+    | Ok { rev_returns; settable; tail_choices } ->
+        begin
+          (* fail if this settable position is not unique *)
+          match find_settable tail_choices with
+          | Error _all_returns -> ()
+          | Ok _another_settable ->
+              failwith "TODO proper error/warning: ambiguous settable position"
+        end;
+        let k_with_placeholder =
+          k Mutable
+            (List.rev_append rev_returns @@
+             trmc_placeholder ::
+             List.map return tail_choices)
+        in
+        let placeholder_pos = List.length rev_returns in
+        let placeholder_pos_lam = Lconst (Const_base (Const_int placeholder_pos)) in
+     (*
+        ∃k, uₖ = Set(dst.u', _) =>
+            Set(
+                (old_dst.
+                  let block = K(return(u₁), .., Placeholder, .., return(uₙ)) in
+                  old_dst <- block,
+                  u'[Dst(block, k)]),
+                (let block = K(return(u₁), .., Placeholderₖ, .., return(uₙ)) in
+                 u'[Dst(block, k)]; block)
+            )
+     *)
+        let let_block_in body =
+          let block_var = Ident.create_local "block" in
+          Llet(Strict, Pgenval, block_var, k_with_placeholder,
+               body block_var)
+        in
+        let block_dst block_var = { var = block_var; offset = placeholder_pos_lam } in
+        Set {
+          dps = (fun old_dst ->
+            let_block_in @@ fun block_var ->
+            Lsequence(assign_to_dst loc old_dst (Lvar block_var),
+                      settable.dps (block_dst block_var))
+          );
+          direct = (fun () ->
+            let_block_in @@ fun block_var ->
+            Lsequence(settable.dps (block_dst block_var),
+                      Lvar block_var)
+          );
+        }
+
+  and choice_prim ctx prim primargs loc =
+    begin [@warning "-8"] (* see choice *)
+      match prim with
+      (* The important case is the construction case *)
+      | Pmakeblock (tag, flag, shape) ->
+          choice_makeblock ctx (tag, flag, shape) primargs loc
+
+      (* Some primitives have arguments in tail-position *)
+      | (Pidentity | Popaque) as idop ->
+          let [l1] = primargs in
+          lift ctx [l1] (fun [l1] -> Lprim (idop, [l1], loc))
+      | (Psequand | Psequor) as shortcutop ->
+          let [l1; l2] = primargs in
+          lift ctx [l2]
+            (fun [l2] -> Lprim (shortcutop, [l1; l2], loc))
+
+      (* cases we don't handle yet *)
+      | (Prevapply | Pdirapply) ->
+          failwith "TODO: should have been simplified away already"
+
+      | (Pmakearray _ | Pduparray _) ->
+          failwith "TODO: we don't handle array indices as destinations yet"
+
+      | Pduprecord _ ->
+          failwith "TODO"
```

Comment:
Are these handled in later commits, or were they forgotten in the "Remaning TODOs"?

Reply:
Note: your comment decided to me to edit the history to push the final form of this pattern-matching already in the first commit. That should make it easier for other people to review the PR. Thanks!